splicemachine/spliceengine
# DB-12095 Backward-compatible Kryo serialization of SQLTimestamp

## Short Description
Kryo does not serialize SQLTimestamp properly.  Inserts of timestamps on Spark store milliseconds precision, even when microseconds are specified (the max precision we can store is 6 digits after the decimal place).

## Long Description
This issue was found before.  SPLICE-2301 attempted to fix it by using a new SQLTimestamp Kryo encoding which stores microseconds precision, but since Kryo serialization is stored in statistics, the change broke backwards compatibility, and a table with stats on a timestamp column, created before the change, could not be dropped.  DB-8205 backed out the change.

The new solution uses the maximum encoded timestamp value (as a long) that could be stored for the timestamp 9999-12-31 23:59:59.999999999 to tell how the timestamp is encoded.  If the encoded value is less than this, use the old Kryo deserialization.  For new Kryo serialization we add an increment value of 70000000000000000L to the encoded timestamp.  When added to the minimum encoded value in the new encoding: TimestampV3DescriptorSerializer.formatLong(new Timestamp(-1899, 0, 1, 0, 0, 0, 0)) + 70000000000000000L, the result will always be greater than the max old encoded value.  We use this fact to tell whether the encoded value uses the old encoding or the new encoding, and use the proper method upon deserialization.

A similar change is made in SpliceSparkKryoRegistrator, to use a new encoding, but its results are never stored on disk, so we skip adding an increment value to the encoded value.

This change is also reducing the number of fractional digits supported in timestamp literals from 9 (nanoseconds) down to 6 (microseconds).  For example, the following SQL statement will now error out:  

>        VALUES ( TIMESTAMP( '9999-12-31 23:59:59.1234567') );

## How to test

> CREATE TABLE TSFIPS ( FIPS1 TIME, FIPS2 TIMESTAMP );
> 
> INSERT INTO TSFIPS VALUES
> ( TIME( '16:03:00'), TIMESTAMP( '1996-08-24 16:03:00.999999') );
> 
> The output of the following should be:
>    FIPS1   |FIPS2
> //--------------------------------------
> 16:03:00|1996-08-24 16:03:00.999999
> 
> SELECT * FROM TSFIPS;
